### PR TITLE
Daily commit - Jihwan Kim

### PR DESCRIPTION
**Pull request by jeekim2**
Pull request 의 제목을 어떻게 쓰는게 적절할 지 모르겠음.

### CODE REVIEW

**Review by YHsla — DISMISSED** [review]

**Review by Bobzzang — APPROVED** [review]
간략하게 문제를 첫 시작 부분에 comment 형태로 넣는건 어떨지 의견 문의 드립니다.
(백준에서 직접 찾아넣는 것도 방법이지만, 문제를 간략하게 작성하면 작성 code를 Review할때 도움이 될 것 같습니다.)

**Comment by jeekim2** [comment]
내껀 왜 리뷰 안해줌 근데?

**Review by Bobzzang — DISMISSED** [review]

File: Problems/3036/ans_3036.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    circles = list(map(int, input().split()))
+    main_c = circles.pop(0)
+
+    ans = []
+    for sub_c in circles:
+        # 유클리드 호제법을 이용한 최대공약수 계산
+        p = main_c
+        q = sub_c
+        while True:
+            p, q = q, p % q
```
[the remark this answers is not in the document]
Comment (jeekim2) [reply Problems/3036/ans_3036.py:18]: circles는 integer list
for 문을 이용해서 iterator 뽑았으니 sub_c는 circles의 element를 순차적으로 가져옴, 즉 integer임

File: Problems/2981/ans_2981.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    # N2 = M * x2 + d
+    # N1 - N2 = M(x1-x2)
+
+    N_list.sort()
+    q = N_list[1] - N_list[0]
+    for i in range(2, num_N):
+        p = N_list[i] - N_list[0]
+        # p should be always bigger than q
+        while True:
+            x = p // q
```
Comment (YHsla) [inline Problems/2981/ans_2981.py:23]: x를 쓰는데가 있는건가요??

**Review by YHsla — COMMENTED** [review]
1st review

File: Problems/2981/ans_2981.py (hunk 2 above)
Comment (jeekim2) [reply Problems/2981/ans_2981.py:23]: > x를 쓰는데가 있는건가요??

깜빡하고 안지운듯?ㅋㅋㅋㅋㅋㅋㅋ

File: Problems/3036/ans_3036.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            if q == 0:
+                break
+        ans.append([main_c // p, sub_c // p])
+
+    for a in ans:
+        print("/".join(list(map(str, a))))
```
[the remark this answers is not in the document]
Comment (jeekim2) [reply Problems/3036/ans_3036.py:24]: > 문자열 포맷팅 이용하면 반복문이 필요 없지 않을까 싶은 생각(?)

난 저런 문법이 어렵드라고... 좋은 방법임.

File: Problems/11399/ans_11399.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    # input = sys.stdin.readline
+    N = int(input())
+    t = list(map(int, input().split()))
+    t.sort()
+    sum = 0
+    w = 0
+    for i in t:
+        w += i
+        sum += w
+    print(sum)
```
[the remark this answers is not in the document]
Comment (jeekim2) [reply Problems/11399/ans_11399.py:14]: 곱하기 vs 더하기
시간 확인 해보면 좋을듯

File: Problems/11050/ans_11050.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def Binomial(n, r):
+    if r < n // 2:
+        r = n - r
+    if n == r:
+        return 1
+    if n - 1 == r:
+        return n
+    return Binomial(n - 1, r) + Binomial(n - 1, r - 1)
```
Comment (Bobzzang) [inline Problems/11050/ans_11050.py:12]: 이렇게 재귀함수를 쓸수가 있네요..

Factorial을 For Loop으로 계산해서 실제 계산을 넣었는데...이게 훨씬 신박하고 깔끔하네요..

질문은 재귀함수를 쓰게 되면 장점은 코드의 단순화랑 속도상에 차이가 있는걸까요?

Comment (Bobzzang) [reply Problems/11050/ans_11050.py:12]: 백준에서는 For로 짜니까 답은 나오는데 바로 "런타임에러" 발생되네요.
문제 푸는게 아니라면, 재귀 함수 적용하는 코딩은 한방에 생각이 나지 않을거 같은데...

Comment (jeekim2) [reply Problems/11050/ans_11050.py:12]: > 

그래서 연습이 필요. 재귀를 하다가 보면 그마저도 시간초과 발생해서, 줄이는 방법을 고민해야 됨.

File: Problems/9461/ans_9461.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# https://www.acmicpc.net/problem/9461
+
+import sys
+
+
+def padovan(i):
+    # P(n) = P(n-1) + P(n-5) in n >= 6
```
Comment (YHsla) [inline Problems/9461/ans_9461.py:7]: P(n) = P(n-2) + P(n-3) in n >=4 규칙이기도 함.

**Review by YHsla — COMMENTED** [review]
211126 review

File: Problems/15651/ans_15651.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    global M
+    global arr
+    if depth >= M:
+        print(" ".join(str(x) for x in arr))
+        return
+    for i in range(N):
+        i += 1
+        # 좋아보이는 코드는 아닌데, for i in range(1,N+1) 보다 갑자기 이뻐보임
+        arr[depth] = i
+        search(depth + 1)
```
[the remark this answers is not in the document]
Comment (jeekim2) [reply Problems/15651/ans_15651.py:21]: 17번 for문으로부터 호출되는 search 함수로 인해, arr 리스트에 값을 누적시켜나가는데,
이 재귀의 끝은 arr 리스트를 모두 채웠을 때가 되니까 return을 넣는 것.
즉슨, if 문을 타지 않는 (arr를 마저 못채우는, 중간 depth) search 함수는 for문을 모두 탄 후
22번줄에서 종료하면서 연쇄적으로 끝나게 됨.

이게 제대로 만드는게 맞는지는 모르겠는데, 여러 번 해 보면서 생기는 내 요령은
14번 if문과 같이, 
재귀 끝 시점을 해당 재귀함수의 시작부분에 (재귀가 시작하자마자 끝나도록) 만드는게 깔끔한 것 같음.

걍 많이 풀어봐야 하는거 같어.. 필요한 부분만, 추적 가능한 재귀를 쓰는게 요령인듯.
depth가 깊어지면 그만큼 시간과 메모리 폭발이 발생함.

File: Problems/15651/ans_15651.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    return
+
+
+def solve():
+    global N
+    global M
+    global arr
+    input = sys.stdin.readline
+    N, M = map(int, input().split())
```
[the remark this answers is not in the document]
Comment (jeekim2) [reply Problems/15651/ans_15651.py:30]: Boundary condition을 벗어나는 input은 채점 test case에서 주어지지 않으니 배제해도 좋지 않을까?
좋게 보면 safety한 설계지만, 일어날 일 없는 input에 리소스를 낭비하는건 과다 설계라는 생각.

File: Problems/9461/ans_9461.py (hunk 6 above)
Comment (jeekim2) [reply Problems/9461/ans_9461.py:7]: 굳
못찾겠던데 ㅋㅋㅋㅋㅋ

File: Problems/2798/ans_2798.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+# https://www.acmicpc.net/problem/2798
+
+
+# 이런 식으로 풀면 안되는데 오늘은 술을 먹어야한다.
```
Comment (YHsla) [inline Problems/2798/ans_2798.py:4]: 브루트포스 알고리즘이 결국 일단 돌리면서 생각하자 방식이니깐 이런식의 풀이가 맞는 것 같은데..
어떤 면에서 불만족하는거에요?

**Review by YHsla — APPROVED** [review]

File: Problems/11051/ans_11051.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+                # 이부분 머리아파서 정리가 안됨... 걍 빡코딩
+
+                try:
+                    t1 = nCr[(i - 1, j - 1)]
+                except:
+                    t1 = nCr[(i - 1, i - j)]
+                try:
+                    t2 = nCr[(i - 1, j)]
+                except:
+                    t2 = nCr[(i - 1, i - j - 1)]
```
Comment (Bobzzang) [inline Problems/11051/ans_11051.py:36]: 특별하게 Dictionary를 사용하는 이유가 있었는지 궁금합니다.

List를 통한 방법이 좀더 심플하지 않을까 싶어서요...

```suggestion
    # 팩토리얼보다 파스칼의 삼각형 이용(이항 계수를 삼각형 모양으로 배열한 것)
    dp = []
    for i in range(N+1):
        dp.append([1]*(i+1))
        print(dp)
    
    for i in range(2, N+1):
        for j in range(1, i):
            t1 = dp[i-1][j-1]
            t2 = dp[i-1][j]
            dp[i][j] = t1 + t2
```

File: Problems/11051/ans_11051.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+                nCr[(i, j)] = (t1 + t2) % DIV
+
+    print(nCr[(N, K)] % DIV)
```
Comment (Bobzzang) [inline Problems/11051/ans_11051.py:39]: 아래 code에서 % DIV를 두번하고 있습니다.
실제로 10007이기 때문에 결과에 영향성이 없지만, 
원래 목적대로 라고 한다면 nCr[(i,j)]는 % DIV를 제거하는 것이 맞다고 보여집니다.

```suggestion
                nCr[(i, j)] = (t1 + t2)

    print(nCr[(N, K)] % DIV)
```

**Review by Bobzzang — APPROVED** [review]
문제 11051 Comment 확인 부탁 드립니다.
문제 의도에 맞기 위해서는 수정이 필요하다고 생각되나, 결과에 영향성은 없으므로 Approve합니다.

File: Problems/11051/ans_11051.py (hunk 10 above)
Comment (jeekim2) [reply Problems/11051/ans_11051.py:36]: 2차원배열로 만들 경우 반복문 돌리면서 쌓아나가게 될 텐데,
각 row마다 길이가 다른 리스트가 들어가는건 좋은 방법이 아니라는 생각이 들었음.
(깔끔해지려면, 정확히 사각형의 matrix가 나와야하지 않을까..?)
더불어 이 문제에선 상관없긴한데, 로직 돌아가면서 1씩 길이가 증가하는 것이 아니라,
갑자기 jump up 해서 memoization을 해야한다면 (4C2 까지밖에 적어놓은것이 없는데, 6C4 같은게 요구된다면)
리스트의 index = 이항계수 라는 공식이 깨져버리게 되는 위험도 있어보임.

File: Problems/11051/ans_11051.py (hunk 11 above)
Comment (jeekim2) [reply Problems/11051/ans_11051.py:39]: 의도된 설계임. 이항계수의 값은 팩토리얼에 기초하므로 조금만 계수가 커져도 큰 수의 계산을 하게 되는데,
어차피 필요한 것은 10007로 나눈 나머지이고 상기 계산방법에서 더하기만을 하고 있으므로
나머지만 딕셔너리에 적음으로써 데이터 양을 줄이려 한 것.
(a + b) % DIV = (a % DIV + b % DIV) % DIV
상기 식은 a,b,DIV가 양의 정수일 때 항상 성립함.

**Review by jeekim2 — COMMENTED** [review]
드디어 merge를 할수있다

**Review by Bobzzang — DISMISSED** [review]
Problem 4948 is reviewed, and approve the coding wrt given problem.

**Review by Bobzzang — DISMISSED** [review]
+= 연산자는 많이 사용했는데, /=는 새롭네요.

코드 Review는 완료하였고, 이견 없어 Approve 합니다.

File: Problems/11653/ans_11653.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# https://www.acmicpc.net/problem/11653
+
+import sys
+
+
+def solve():
+    input = sys.stdin.readline
+    N = int(input())
+    div = 2
+    while N != 1:
```
Comment (YHsla) [inline Problems/11653/ans_11653.py:10]: N=1일 때는 출력이 없는 것 아닌가..?

**Review by YHsla — APPROVED** [review]

File: Problems/11051/ans_11051.py (hunk 11 above)
Comment (jeekim2) [reply Problems/11051/ans_11051.py:39]: > Comment에 대한 설명 감사합니다. 그런 의도가 있을거라고는 생각하지 못했네요.
>근데 for loop 동안 나누기를 계속 수행하는데 큰 수를 나누더라도 마지막에 1회 하는 것이 runtime 시간이 더 짧지 않을까요? 어떤 계산 방식이나 알고리즘이 코드의 속도에 영향을 주는지 스터디 해주는 것도 좋을거 같네요ㅎㅎㅎ

Overflow에 대해 자유로운 language가 아닌 language보다 적다는걸 생각해봤을 때,
큰 수 계산을 신경 안쓰고 알고리즘을 공부하는건 좋은 어프로치가 아니라는 개인적인 생각에 따라...
파이썬 상에서 효율적인 코드보다는 범용적으로 효율적인 코드를 만드는 어프로치를 하고싶음.

File: Problems/2798/ans_2798.py (hunk 9 above)
Comment (jeekim2) [reply Problems/2798/ans_2798.py]: > 

for문 무지성 중첩하는게 별로 보기 싫어서 ㅋㅋㅋㅋㅋㅋ

File: Problems/11653/ans_11653.py (hunk 12 above)
Comment (Bobzzang) [reply Problems/11653/ans_11653.py:10]: 인수분해이기 때문에, 나누기를 계속하면서 나누기 값이 1이 되면 완료되는 구조라고 이해됩니다.

**Comment by jeekim2** [comment]
그래서 내껀 누가 리뷰해주나..

File: Problems/2231/ans_2231.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+import sys
+
+
+def solve():
+    input = sys.stdin.readline
+    N = int(input())
+
+    digit = len(str(N))
+
+    startNum = N - (digit * 9)
```
Comment (Bobzzang) [inline Problems/2231/ans_2231.py:12]: 이 부분을 이해하는데 좀 시간이 걸렸네요.
N의 자릿수 * 9를 통해서 최소 시작 숫자를 찾는건 좋은 방법인거 같습니다.

File: Problems/2798/ans_2798_KJH.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+    max_sum = 0
+    for i in cards:
+        for j in cards:
+            if j != i:
+                for k in cards:
+                    if k != j and k != i:
+                        temp_sum = i + j + k
+                        if max_sum < temp_sum and temp_sum <= M:
+                            max_sum = temp_sum
```
Comment (Bobzzang) [inline Problems/2798/ans_2798_KJH.py:22]: 이건 카드 index 정보를 사용하는게 더 나을거 같네요.
(영한이가 동일 문제 풀었던 방식에서 index 방식으로 i,j,k가 동일한 카드일 경우에 대한 고려를 넣었음.)

File: Problems/2231/ans_2231.py (hunk 13 above)
Comment (YHsla) [reply Problems/2231/ans_2231.py:12]: 나도 좋은 아이디어라고 생각됨.
결국 반복 횟수를 줄이는 방향의 아이디어라고 한다면 N의자릿수*9를 빼서 최소점부터 찾는 것보다 
범위가 0~N의 자릿수*9 사이가 될테니 그 중간 middle을 기준으로 오름 또는 내림 방향으로 확인하는 방식으로 접근해도 좋지 않을까 싶음. (그냥 머릿속으로만 생각해본 방향...)

**Review by YHsla — APPROVED** [review]

File: Problems/11653/ans_11653.py (hunk 12 above)
Comment (YHsla) [reply Problems/11653/ans_11653.py:10]: N==1이되면 while을 빠져나오는 건 알겠는데... print로 출력이 되남요?

File: Problems/2231/ans_2231.py (hunk 13 above)
Comment (jeekim2) [reply Problems/2231/ans_2231.py:12]: 뭐지 approve 한거 아닌가? 왜 안뜨지..

Comment (hongmyeong-lee) [reply Problems/2231/ans_2231.py:12]: N자리수*9를 통해서 풀이를 한것에 대한 설명이 좀더 있었으면 좋았을 거 같습니다.
startnum는 결국 예를 들어 3자리라고 가정했을 경우 그 각각의 자리수가 999인 경우가 최대일텐데 그렇게 하더라도 10의 자리, 100의 자리를 더해서 나오는 값이 있기 때문에 startnum부터 시작해도 생성자를 구할 수 있어 그렇게 한거 같은데 제가 이해한게 맞는지 모르겠네요
밑에는 pytest 돌릴려고 만든거 같은데 공부좀해보겠습니다

**Review by hongmyeong-lee — APPROVED** [review]

File: Problems/2231/ans_2231.py (hunk 13 above)
Comment (jeekim2) [reply Problems/2231/ans_2231.py:12]: 설명이 부족했던건 맞는데, 어디까지 설명을 달아야하느냐가 항상 문제인듯.
